src/main/kotlin/org/partiql/scribe/sql/SqlDialect.kt

```diff
@@ -0,0 +1,852 @@
+        var t = tail
+        t = visitExprWrapped(node.expr, t)
+        t = t concat " AS ${node.asAlias.sql()}"
+        return t
+    }
+
+    // GROUP BY
+
+    override fun visitGroupBy(node: GroupBy, tail: SqlBlock): SqlBlock {
+        var t = tail
+        t = t concat when (node.strategy) {
+            GroupBy.Strategy.FULL -> "GROUP BY "
+            GroupBy.Strategy.PARTIAL -> "GROUP PARTIAL BY "
+        }
+        t = t concat list("", "") { node.keys }
+        t = if (node.asAlias != null) t concat " GROUP AS ${node.asAlias!!.sql()}" else t
+        return t
+    }
+
+    override fun visitGroupByKey(node: GroupBy.Key, tail: SqlBlock): SqlBlock {
+        var t = tail
+        t = visitExprWrapped(node.expr, t)
+        t = if (node.asAlias != null) t concat " AS ${node.asAlias!!.sql()}" else t
+        return t
+    }
+
+    // SET OPERATORS
+
+    override fun visitSetOp(node: SetOp, tail: SqlBlock): SqlBlock {
+        val op = when (node.setq) {
+            null -> node.type.name
+            else -> "${node.type.name} ${node.setq!!.name}"
+        }
+        return tail concat op
+    }
+
+    override fun visitExprSFWSetOp(node: Expr.SFW.SetOp, tail: SqlBlock): SqlBlock {
+        var t = tail
+        t = visitSetOp(node.type, t)
+        t = t concat SqlBlock.Nest(
+            prefix = " (",
+            postfix = ")",
+            child = SqlBlock.root().apply { visitExprSFW(node.operand, this) },
+        )
+        return t
+    }
+
+    // ORDER BY
+
+    override fun visitOrderBy(node: OrderBy, tail: SqlBlock): SqlBlock =
+        tail concat list("ORDER BY ", "") { node.sorts }
+
+    override fun visitSort(node: Sort, tail: SqlBlock): SqlBlock {
+        var t = tail
+        t = visitExprWrapped(node.expr, t)
+        t = when (node.dir) {
+            Sort.Dir.ASC -> t concat " ASC"
+            Sort.Dir.DESC -> t concat " DESC"
+            null -> t
+        }
+        t = when (node.nulls) {
+            Sort.Nulls.FIRST -> t concat " NULLS FIRST"
+            Sort.Nulls.LAST -> t concat " NULLS LAST"
+            null -> t
+        }
+        return t
+    }
+
+    // --- Block Constructor Helpers
+
+    private infix fun SqlBlock.concat(rhs: String): SqlBlock {
+        next = SqlBlock.Text(rhs)
+        return next!!
+    }
+
+    private infix fun SqlBlock.concat(rhs: SqlBlock): SqlBlock {
+        next = rhs
+        return next!!
+    }
+
+    private fun type(symbol: String, vararg args: Int?, gap: Boolean = false): SqlBlock {
+        val p = args.filterNotNull()
+        val t = when {
+            p.isEmpty() -> symbol
+            else -> {
+                val a = p.joinToString(",")
+                when (gap) {
+                    true -> "$symbol ($a)"
+                    else -> "$symbol($a)"
+                }
+            }
+        }
+        // types are modeled as text; as we don't way to reflow
+        return SqlBlock.Text(t)
+    }
+
+    private fun list(
```

Comment:
nit: the other dialects currently have their own private versions of `list`. could create an `internal`/`public` version of this utility function that could be reused by all of the dialects